src/backend/src/lib.rs

```diff
@@ -1,33 +1,143 @@
+};
 use k256::PublicKey;
 use serde_bytes::ByteBuf;
 use shared::http::{HttpRequest, HttpResponse};
 use shared::metrics::get_metrics;
 use shared::std_canister_status;
+use std::borrow::Cow;
 use std::cell::RefCell;
 use std::str::FromStr;
 
 mod guards;
 
+type VMem = VirtualMemory<DefaultMemoryImpl>;
+type ConfigCell = StableCell<Option<Candid<Config>>, VMem>;
+type ChainId = u64;
+type UserTokenMap = StableBTreeMap<StoredPrincipal, Candid<Vec<Token>>, VMem>;
+
+const CONFIG_MEMORY_ID: MemoryId = MemoryId::new(0);
+const USER_TOKEN_MEMORY_ID: MemoryId = MemoryId::new(1);
+
+const MAX_SYMBOL_LENGTH: usize = 20;
+const MAX_TOKEN_LIST_LENGTH: usize = 100;
+
 thread_local! {
-    static STATE: RefCell<Option<State>> = RefCell::default();
+    static MEMORY_MANAGER: RefCell<MemoryManager<DefaultMemoryImpl>> = RefCell::new(
+        MemoryManager::init(DefaultMemoryImpl::default())
+    );
+
+    static STATE: RefCell<State> = RefCell::new(
+        MEMORY_MANAGER.with(|mm| State {
+            config: ConfigCell::init(mm.borrow().get(CONFIG_MEMORY_ID), None).expect("config cell initialization should succeed"),
+            user_token: UserTokenMap::init(mm.borrow().get(USER_TOKEN_MEMORY_ID)),
+        })
+    );
 }
 
 pub fn read_state<R>(f: impl FnOnce(&State) -> R) -> R {
-    STATE.with(|cell| f(cell.borrow().as_ref().expect("state not initialized")))
+    STATE.with(|cell| f(&*cell.borrow()))
+}
+
+pub fn mutate_state<R>(f: impl FnOnce(&mut State) -> R) -> R {
+    STATE.with(|cell| f(&mut *cell.borrow_mut()))
+}
+
+pub fn read_config<R>(f: impl FnOnce(&Config) -> R) -> R {
+    read_state(|state| {
+        f(state
+            .config
+            .get()
+            .as_ref()
+            .expect("config is not initialized"))
+    })
+}
+
+#[derive(Default)]
+struct Candid<T>(T)
+where
+    T: CandidType + for<'de> Deserialize<'de>;
+
+impl<T> Storable for Candid<T>
+where
+    T: CandidType + for<'de> Deserialize<'de>,
+{
+    const BOUND: Bound = Bound::Unbounded;
+
+    fn to_bytes(&self) -> Cow<'_, [u8]> {
+        Cow::Owned(candid::encode_one(&self.0).expect("encoding should always succeed"))
+    }
+
+    fn from_bytes(bytes: Cow<'_, [u8]>) -> Self {
+        Self(candid::decode_one(bytes.as_ref()).expect("decoding should succeed"))
+    }
+}
+
+impl<T> Deref for Candid<T>
+where
+    T: CandidType + for<'de> Deserialize<'de>,
+{
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.0
+    }
 }
 
-#[derive(CandidType, Deserialize)]
 pub struct State {
+    config: ConfigCell,
+    user_token: UserTokenMap,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct Token {
+    pub contract_address: String,
+    pub chain_id: ChainId,
+    pub symbol: Option<String>,
+    pub decimals: Option<u8>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+struct StoredPrincipal(Principal);
+
+impl Storable for StoredPrincipal {
+    const BOUND: Bound = Bound::Bounded {
+        max_size: 29,
```

Comment:
it's not 32 bytes for a Principal?

Reply:
I know that account identifiers are typically 32 bytes long.  This test passes:
```
/// We assume that new account identifiers are 32 bytes long.
///
/// Note: Old-style account identifiers were 28 bytes long.
#[test]
fn account_identifier_has_32_bytes() {
    let account_identifier = icp_ledger::AccountIdentifier::from(ic_base_types::PrincipalId::new_user_test_id(99));
    assert_eq!(
        account_identifier.to_vec().len(),
        AccountStorageKey::ACCOUNT_IDENTIFIER_MAX_BYTES // Note: This is 32.
    );
}
```

Reply:
I would check the spec and the code that defines principal to be sure though.

Reply:
The [candid code that defines Principal](https://docs.rs/candid/0.9.10/src/candid/types/principal.rs.html#85) says 29 bytes and mentions a 4 byte CRC.

Reply:
By the way, I just got a notification from GitHub saying that a review was requested.  I guess that given that this is OISY wallet, the review is probably wanted from @peterpeterparker but I am happy to pitch in as well, if wanted.  I don't know the business context but can read Rust, if wanted.  Please let me know.

Reply:
Yeah nvm my comment, my brain was off for a minute.